src/libraries/System.Private.CoreLib/src/System/IO/PathInternal.Windows.cs

```diff
@@ -409,5 +410,34 @@ internal static bool IsEffectivelyEmpty(ReadOnlySpan<char> path)
             }
             return true;
         }
+
+        // this method works only for `fullPath` returned by Path.GetFullPath
+        // currently we don't have interest in supporting relative paths
+        internal static void DosToNtPath(ReadOnlySpan<char> fullPath, ref ValueStringBuilder vsb)
```

Comment:
This is attempting to emulate the behavior of RtlDosPathNameToRelativeNtPathName_U_WithStatus?  Why are we not just calling that?

These are the kinds of issues I was concerned about with switching from CreateFile to NtCreateFile.

Reply:
> Why are we not just calling that?

Because this particular method is internal and not documented at all.

> These are the kinds of issues I was concerned about with switching from CreateFile to NtCreateFile.

With this PR we should have all cases covered (devices `\\.\` and UNC `\\`)

Reply:
> With this PR we should have all cases covered

Famous last words ;-)

Remind me why we can't just keep using CreateFile?

Reply:
> Remind me why we can't just keep using CreateFile?

We can use it for cases where `preallocationSize` was not specified (this is something supported only by `NtCreateFile`).

I was hoping that we could always use `NtCreateFile` to have the same behavior no matter if `preallocationSize` was provided or not. And also slightly better perf (fewer syscalls).

Reply:
And the atomic preallocationSize feature (which is already non-atomic on Unix) is worth it?

Reply:
> And the atomic preallocationSize feature (which is already non-atomic on Unix) is worth it?

I think that it's definitely worth trying, but the timing (close to release) is very unfortunate.

We would have definitely avoided that if we had tests for UNCs and devices. I've filled #54495 to keep track of it and hopefully address it in .NET 7.

Reply:
> the timing (close to release) is very unfortunate

Right.  Correct me if I'm wrong, but we've now seen at least three issues related to the switch, and I fully expect we'll see more.  My suggestion is to go back to CreateFile for this release, and if that means not supporting an atomic preallocationSize, we remove that for the release as well.

Reply:
> but we've now seen at least three issues related to the switch

You are right, however they are all being solved by this PR:
* #54474 and #54475 is the same issue: lack of handling of the UNC paths (`\??\UNC` for files located on remote machines)
* #54337: lack of handling of device paths (`\\.\`)

Reply:
If you're pushing to keep it, then #54495 really needs to be in .NET 6, not .NET 7.